server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalMedianAbsoluteDeviation.java

```diff
@@ -0,0 +1,117 @@
+
+    protected final TDigestState valuesSketch;
+
+    public InternalMedianAbsoluteDeviation(String name,
+                                           List<PipelineAggregator> pipelineAggregators,
+                                           Map<String, Object> metaData,
+                                           DocValueFormat format,
+                                           TDigestState valuesSketch) {
+
+        super(name, pipelineAggregators, metaData);
+        this.format = format;
+        this.valuesSketch = valuesSketch;
+    }
+
+    public InternalMedianAbsoluteDeviation(StreamInput in) throws IOException {
+        super(in);
+        format = in.readNamedWriteable(DocValueFormat.class);
+        valuesSketch = TDigestState.read(in);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(format);
+        TDigestState.write(valuesSketch, out);
+    }
+
+    @Override
+    public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        TDigestState valueMerged = null;
+        for (InternalAggregation aggregation : aggregations) {
+            final InternalMedianAbsoluteDeviation magAgg = (InternalMedianAbsoluteDeviation) aggregation;
+            if (valueMerged == null) {
+                valueMerged = new TDigestState(magAgg.valuesSketch.compression());
+            }
+            valueMerged.add(magAgg.valuesSketch);
+        }
+
+        return new InternalMedianAbsoluteDeviation(name, pipelineAggregators(), metaData, format, valueMerged);
+    }
+
+    @Override
+    public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        final boolean anyResults = valuesSketch.size() > 0;
+        final Double mad = anyResults
+            ? getMedianAbsoluteDeviation()
+            : null;
+
+        builder.field(CommonFields.VALUE.getPreferredName(), mad);
+        if (format != DocValueFormat.RAW && anyResults) {
+            builder.field(CommonFields.VALUE_AS_STRING.getPreferredName(), format.format(mad).toString());
+        }
+
+        return builder;
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(valuesSketch);
+    }
+
+    @Override
+    protected boolean doEquals(Object obj) {
+        InternalMedianAbsoluteDeviation other = (InternalMedianAbsoluteDeviation) obj;
+        return Objects.equals(valuesSketch, other.valuesSketch);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return MedianAbsoluteDeviationAggregationBuilder.NAME;
+    }
+
+    @Override
+    public double value() {
+        return getMedianAbsoluteDeviation();
+    }
+
+    @Override
+    public double getMedianAbsoluteDeviation() {
+        return computeMedianAbsoluteDeviation(valuesSketch);
```

Comment:
This is a relatively expensive operation, I think we should calculate this value once and store it rather than calculating it each time the getter is called, especially as the value cannot change. Maybe we should calculate this value in the constructor and store it along side the sketch or calculate it in the reduce method and store it alongside the sketch at that point?

Reply:
Definitely agree, I had a couple ways of doing it before but nixed them because they were gross. Don't know why doing it in the constructor didn't occur to me, that's much cleaner